### PR TITLE
feat(): added {Event,Action}[Store] and registered stores

### LINKED ISSUE #16: Actions System

### PR DESCRIPTION
**Pull request by kyranet**
Also did some fixes on Client and added missing member accessors.

<details>
<summary>Proof of Concept</summary>

```ts
import { Action } from '../lib/structures/Action';
import type { Structure, ActionStore } from '..';
import type { MessageCreateDispatch } from '@klasa/ws';

export default class ActionMessageCreate extends Action {

    public constructor(store: ActionStore, directory: string, file: readonly string[]) {
        super(store, directory, file, { event: 'MESSAGE_CREATE' });
    }

    public check(data: MessageCreateDispatch): Structure | null {
        console.log(data);
        return null;
    }

    public build(): Structure {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return null!;
    }

    public cache(): void {
        // data.channel.messages._add(data);
    }

}
```

**Results on:**

image

---

</details>

Fixes #16.

### CODE REVIEW

File: src/client/Client.ts [hunk 1]
```diff
@@ -7,10 +7,8 @@ import { ClientOptionsDefaults } from '../util/Constants';
 
 import type { Store } from '../lib/structures/base/Store';
 import type { Piece } from '../lib/structures/base/Piece';
-
-export interface Plugin {
-	[Client.plugin]: Function;
-}
+import { EventStore } from '../lib/structures/EventStore';
+import { ActionStore } from '../lib/structures/ActionStore';
```
Comment (vladfrangu) [inline src/client/Client.ts:11]: Move above type imports

File: src/lib/structures/Action.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import type { DispatchPayload } from '@klasa/ws';
+import type { Structure } from '../../client/caching/structures/base/Structure';
+
+/**
+ * The common class for all actions.
+ */
+export abstract class Action<T extends DispatchPayload = DispatchPayload, S extends Structure = Structure> extends Event {
+
+	/**
+	 * The name of the event from {@link Client} to be fired.
```
Comment (vladfrangu) [inline src/lib/structures/Action.ts:15]:
```suggestion
	 * The name of the event that the {@link Client} will fire.
```

File: src/lib/structures/Action.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+		const built = this.build(data);
+		if (built) {
+			this.cache(built);
+			this.client.emit(this.clientEvent, built);
+		}
+	}
+
+	/**
+	 * Checks whether or not the data structure was already cached, returning it when it does.
```
Comment (vladfrangu) [inline src/lib/structures/Action.ts:59]:
```suggestion
	 * Checks whether or not the data structure was already cached, returning it if it was.
```

File: src/lib/structures/ActionStore.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	/**
+	 * Clears the events from the store and removes the listeners.
+	 * @since 0.0.1
+	 */
+	public clear(): void {
+		for (const event of this.values()) this.delete(event);
+	}
+
+	/**
+	 * Deletes an event from the store.
```
Comment (vladfrangu) [inline src/lib/structures/ActionStore.ts:29]:
```suggestion
	 * Deletes an action from the store.
```

File: src/lib/structures/ActionStore.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	 * @since 0.0.1
+	 * @param name An event object or a string representing the event name.
```
Comment (vladfrangu) [inline src/lib/structures/ActionStore.ts:31]:
```suggestion
	 * @param name An action object or a string representing the action name.
```

File: src/lib/structures/ActionStore.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	 * @param client The client this Store was created with
+	 */
+	public constructor(client: Client) {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+		// @ts-ignore
+		super(client, 'actions', Action);
+	}
+
+	/**
+	 * Clears the events from the store and removes the listeners.
```
Comment (vladfrangu) [inline src/lib/structures/ActionStore.ts:21]:
```suggestion
	 * Clears the actions from the store and removes the listeners.
```

File: src/lib/structures/ActionStore.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	public delete(name: Action | string): boolean {
+		const event = this.resolve(name);
+		if (!event) return false;
+		// eslint-disable-next-line dot-notation
+		event['_unlisten']();
+		return super.delete(event);
+	}
+
+	/**
+	 * Sets up an event in our store.
```
Comment (vladfrangu) [inline src/lib/structures/ActionStore.ts:43]:
```suggestion
	 * Sets up an action in our store.
```

File: src/lib/structures/Event.ts [hunk 8]
```diff
@@ -0,0 +1,150 @@
+import { Piece, PieceOptions } from './base/Piece';
+
+import type { EventStore } from './EventStore';
+import type { EventEmitter } from 'events';
+import type { Client } from '../../client/Client';
```
Comment (vladfrangu) [inline src/lib/structures/Event.ts:5]:
```suggestion
import type { EventEmitter } from 'events';

import type { EventStore } from './EventStore';
import type { Client } from '../../client/Client';
```

File: src/lib/structures/Event.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+			...super.toJSON(),
+			once: this.once,
+			event: this.event,
+			emitter: this.emitter.constructor.name
+		};
+	}
+
+}
+
+export interface Event {
```
Comment (vladfrangu) [inline src/lib/structures/Event.ts:128]: Unrelated to review, what is this for?

File: src/lib/structures/EventStore.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	 * @since 0.0.1
+	 */
+	private readonly _onceEvents = new Set<string>();
+
+	/**
+	 * @since 0.0.1
+	 * @param client The client this Store was created with
+	 */
+	public constructor(client: Client) {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
```
Comment (vladfrangu) [inline src/lib/structures/EventStore.ts]: Ditto as actions per Discord question

File: src/lib/structures/Event.ts (hunk 9 above)
Comment (kyranet) [reply src/lib/structures/Event.ts:129]: Augments `Event` so `Event#store` is an `EventStore` instance, giving access to `this.store['_onceEvents']` (line 94).

**Pull request merged**